### ISSUE TITLE
reinterpret cast between long* and Int64* breaks strict aliasing rules

### ISSUE THREAD

**Issue by mgaunard** [bloomberg/bde#189]
Int64 is defined as being __int64 on Windows and long long elsewhere.
Yet some code is casting long* on 64-bit to Int64* and accessing data, which breaks the strict aliasing rules.

Examples include uninitializedFillN.

**Comment by mversche** [comment]
File name, line number?  

(I just looked at https://github.com/bloomberg/bde/blob/2a3a906948bbd414ad4dfc32e16de27e48b3db2e/groups/bsl/bslalg/bslalg_arrayprimitives.cpp and it wasn't obvious to me)

**Comment by abeels** [comment]
- bslalg_arrayprimitives.h:2633 and bslalg_arrayprimitives.h:2657

These two lines are an out-and-out bug, because `sizeof(long) != sizeof(Int64)` on our supported platforms.

Also, atomic operations code violates strict aliasing on powerpc:

- Any use of `Int64_Words::value` in bsls_atomicoperations_powerpc32_aix_xlc.cpp, such as bsls_atomicoperations_powerpc32_aix_xlc.cpp:75, 95, and 120.

**Comment by mversche** [comment]
bslalg_arrayprimitives.h:2633 and bslalg_arrayprimitives.h:2657

The `#ifdef` branch might not be the best choice for implementing this logic, but it is correct -- long in 64bits on most unix platforms when performing a 64bit build.

`bsls_atomicoperations_powerpc32_aix_xlc.cpp` does have a questionable use of type aliasing, but it is intertwined with the inline assembly instructions used to implement the 64bit operations using a 32bit instruction set.  Could it be pulled apart? Probably.  Should we? I'm skeptical.

**Comment by abeels** [comment]
Yes, I got myself confused about `sizeof(long)`.  It's unstable, but the correct size in this case.

I'm not suggesting that we re-write the assembler in `bsls_atomicoperations`.  I am just continuing the search for code that we might have to re-evaluate if we want to get rid of our aliasing flags on all platforms.
